src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -391,7 +391,9 @@ void PipelineExecutor::executeImpl(size_t num_threads, bool concurrency_control)
     SCOPE_EXIT_SAFE(
         if (!finished_flag)
         {
-            finish();
+            /// If finished_flag is not set, there was an exception.
+            /// Cancel execution in this case.
+            cancel();
```

Comment:
`cancel()` actually calls `finish()` inside. Also, unclear if there is any reason to call only `finish()` here